components/concept/storage/src/main/java/mozilla/components/concept/storage/BookmarksStorage.kt

```diff
@@ -0,0 +1,116 @@
+     * @param recursive Whether to recurse and obtain all levels of children.
+     * @return The populated root starting from the guid.
+     */
+    suspend fun getTree(guid: String, recursive: Boolean = false): BookmarkNode?
+
+    /**
+     * Searches bookmarks with a query string.
+     *
+     * @param query The query string to search.
+     * @param limit The maximum number of entries to return.
+     * @return The list of matching bookmark nodes up to the limit number of items.
+     */
+    suspend fun searchBookmarks(query: String, limit: Int = defaultBookmarkSearchLimit): List<BookmarkNode>
+
+    /**
+     * Adds a new bookmark item to a given node.
+     *
+     * Sync behavior: will add new bookmark item to remote devices.
+     *
+     * @param parentGuid The parent guid of the new node.
+     * @param newNode The details of the node to add.
+     * @param position The optional position to add the new node or null to append.
+     * @return The guid of the newly inserted bookmark item.
+     */
+    suspend fun addItem(parentGuid: String, newNode: BookmarkNode, position: Int? = null): String
+
+    /**
+     * Adds a new bookmark folder to a given node.
+     *
+     * Sync behavior: will add new separator to remote devices.
+     *
+     * @param parentGuid The parent guid of the new node.
+     * @param newNode The details of the node to add.
+     * @param position The optional position to add the new node or null to append.
+     * @return The guid of the newly inserted bookmark item.
+     */
+    suspend fun addFolder(parentGuid: String, newNode: BookmarkNode, position: Int? = null): String
+
+    /**
+     * Adds a new bookmark separator to a given node.
+     *
+     * Sync behavior: will add new separator to remote devices.
+     *
+     * @param parentGuid The parent guid of the new node.
+     * @param newNode The details of the node to add.
+     * @param position The optional position to add the new node or null to append.
+     * @return The guid of the newly inserted bookmark item.
+     */
+    suspend fun addSeparator(parentGuid: String, newNode: BookmarkNode, position: Int? = null): String
+
+    /**
+     * Edits the properties of an existing bookmark item and/or moves an existing one underneath a new parent guid.
+     *
+     * Sync behavior: will alter bookmark item on remote devices.
+     *
+     * @param guid The guid of the item to update.
+     * @param info The info to change in the bookmark.
+     */
+    suspend fun updateNode(guid: String, info: BookmarkInfo)
+
+    /**
+     * Deletes a bookmark node and all of its children, if any.
+     *
+     * Sync behavior: will remove bookmark from remote devices.
+     *
+     * @return Whether the bookmark existed or not.
+     */
+    suspend fun deleteNode(guid: String): Boolean
+
+    /**
+     * The default number of bookmark entries to return from searches
+     */
+    val defaultBookmarkSearchLimit: Int
```

Comment:
Ah, i see what you meant here. Yeah, this is worse than before, my bad! Let's just switch back to what you had before. Defining this as a requirement for interface implementors is pretty weird.

Reply:
I agree. Going back to the old solution.